### DISCUSSION TITLE
Advice on network configuration

### DISCUSSION THREAD

**Discussion by nirgivon** [mavlink-router/mavlink-router#406] [Q&A] [answered]
Hello,
I hope to get advice on how to configure Mavlink Router to work in this setup. (See attached file)

I need to have the 2 pilots "see" eath other in order for the Antenna tracker to fallow the plane.
The Mission Planner, using UDP connection, can only connect to one Endpoint.

How should I configure "Mavlink Rounter" on the 2 raspberry pi's so "all" units "see" each other?

![Mavlink Router Network config](https://github.com/mavlink-router/mavlink-router/assets/26519641/ac50b115-2ba3-4443-a5ba-7a475c8ecaee)

**Comment by jbeyerstedt** [comment]
I guess "pilot" is describing the FMU (running ArduPilot, PX4, whatever) and not a person. And I assume, that the two raspberry pis can reach each other via IP based communication and you know their IP address.

The most reliable way would use the TCP endpoint of mavlink-router. One raspberry pi connects to the other one using a `TcpEndpoint` (see [config.sample](https://github.com/mavlink-router/mavlink-router/blob/master/examples/config.sample#L209)). The TCP port is enabled by default and [listens on port 5760](https://github.com/mavlink-router/mavlink-router/blob/master/examples/config.sample#L49). Each of your two mavlink-router instances also needs an `UartEndpoint` of course.

You need to keep in might though, that TCP will re-transmit packages if the wireless link is not reliable. This might not be a desired feature with telemetry data as old data is not relevant any more and therefore does not need re-transmission (and MAVLink will do re-transmissions for important stuff on it's own).  

So you could also use a UDP connection. One mavlink-router config needs to have a `UdpEndpoint` in `server` mode and the other one in `normal` mode. On the server side, you need to [choose a port number for listening](https://github.com/mavlink-router/mavlink-router/blob/master/examples/config.sample#L192) and the "client" side needs to have the IP address and port of the server in its config. It's important, that the mavlink-router instance with the "client" config has some MAVLink system which actively sends heartbeat messages. (This should be the case for any FMU while Q-Groundcontrol usually doesn't send a heartbeat until it has received one from a drone.)

Each mavlink-router instance can then also have other endpoints to connect to other systems on their side. Mavlink-router will figure out who is reachable via which connection then. Only one endpoint is needed to connect the antenna tracker side with the drone side.

**Reply by nirgivon** [reply]
Hi, Thank so much you for the fast and detailed reply!!!
You are correct with your assumption, when I say Pilot, I mean cube Orange in my case.
Both Pi's can "see" each other via the LAN.
and I have full control over the IP scheme.

Do you think it matters if the Server end point is on the drone side or the antenna side?

Should the Mission Planner "connect" to the pi that is in the antenna tracker?
Should the mavlink router on the antenna tracker have 2 endpoints configured? one for communication with the drone and one for mission planner?

**Reply by jbeyerstedt** [reply, marked as answer]
> Do you think it matters if the Server end point is on the drone side or the antenna side?

As mentioned before: When using an UDP connection, it only matters, if side does not send heartbeats on its own. With TCP that issue should not exist. I would put the server side on the antenna tracker, as this should be online first I guess.

> Should the mavlink router on the antenna tracker have 2 endpoints configured? one for communication with the drone and one for mission planner?

I only know QGC, which listens on UDP port 14550 for incoming messages. Assuming, that Mission Planner behaves similarly, you could just add a "normal" mode UDP endpoint with the IP address of your PC for your ground station connection. Alternatively, you could also add a TCP connection to Mission Planner which connects to mavlink-router's TCP port. This is usually more reliable and allows using different devices running the ground station software as mavlink-router does not need to know the PC's IP address. Mavlink-router's TCP server will implicitly add a new endpoint internally for each new client.

> Should the Mission Planner "connect" to the pi that is in the antenna tracker?

Yes. (You could also connect to the one on the drone, but then your wireless link would need to carry telemetry messages twice. Otherwise, mavlink-router on the antenna tracker will take care of distributing messages to all connected systems.)

**Reply by nirgivon** [reply]
Thank you very much!

Great help.
